homeassistant/components/zone/__init__.py

```diff
@@ -95,79 +121,227 @@ def async_active_zone(hass, latitude, longitude, radius=0):
     return closest
 
 
-async def async_setup(hass, config):
+def in_zone(zone: State, latitude: float, longitude: float, radius: float = 0) -> bool:
+    """Test if given latitude, longitude is in given zone.
+
+    Async friendly.
+    """
+    zone_dist = distance(
+        latitude,
+        longitude,
+        zone.attributes[ATTR_LATITUDE],
+        zone.attributes[ATTR_LONGITUDE],
+    )
+
+    if zone_dist is None or zone.attributes[ATTR_RADIUS] is None:
+        return False
+    return zone_dist - radius < cast(float, zone.attributes[ATTR_RADIUS])
+
+
+class ZoneStorageCollection(collection.StorageCollection):
+    """Zone collection stored in storage."""
+
+    CREATE_SCHEMA = vol.Schema(CREATE_FIELDS)
+    UPDATE_SCHEMA = vol.Schema(UPDATE_FIELDS)
+
+    async def _process_create_data(self, data: Dict) -> Dict:
+        """Validate the config is valid."""
+        return cast(Dict, self.CREATE_SCHEMA(data))
+
+    @callback
+    def _get_suggested_id(self, info: Dict) -> str:
+        """Suggest an ID based on the config."""
+        return cast(str, info[CONF_NAME])
+
+    async def _update_data(self, data: dict, update_data: Dict) -> Dict:
+        """Return a new updated data object."""
+        update_data = self.UPDATE_SCHEMA(update_data)
+        return {**data, **update_data}
+
+
+class IDLessCollection(collection.ObservableCollection):
+    """A collection without IDs."""
+
+    counter = 0
+
+    async def async_load(self, data: List[dict]) -> None:
+        """Load the collection. Overrides existing data."""
+        for item_id in list(self.data):
+            await self.notify_change(collection.CHANGE_REMOVED, item_id, None)
```

Comment:
Would this cause issues on reload, if for some reason users would customize entity_id/name through entity registry?

Reply:
They can't as it won't have an ID, as ID is pulled from the config which does not contain `CONF_ID`

Reply:
Nevermind. For some reason I thought they would show up in the entity registry, but I don't see them there.